### PR TITLE
[GISel] LegalizationArtifactCombiner: Elide redundant G_SEXT_INREG

### PR DESCRIPTION
**Pull request by s-barannikov**
This is similar to 373c343a, but for targets with zero-or-negative-one booleans.
The difference in tests is mostly due to G_SEXT_INREG being illegal for some targets, in which case it gets expanded into G_SHL/G_ASHR pair, which is not currently optimized by the combiner.

### CODE REVIEW

File: llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h [hunk 1]
```diff
@@ -211,7 +212,14 @@ class LegalizationArtifactCombiner {
       uint64_t SizeInBits = SrcTy.getScalarSizeInBits();
       if (DstTy != MRI.getType(TruncSrc))
         TruncSrc = Builder.buildAnyExtOrTrunc(DstTy, TruncSrc).getReg(0);
-      Builder.buildSExtInReg(DstReg, TruncSrc, SizeInBits);
+      // Elide G_SEXT_INREG if possible. This is similar to eliding G_AND in
+      // tryCombineZExt. Refer to the comment in tryCombineZExt for rationale.
+      if (KB && KB->computeNumSignBits(TruncSrc) >=
+                    DstTy.getScalarSizeInBits() - SizeInBits + 1)
```
Comment (arsenm) [inline llvm/include/llvm/CodeGen/GlobalISel/LegalizationArtifactCombiner.h:218]: x >= y + 1 -> x > y?

**Comment by aemerson** [comment]
In the original review for https://github.com/llvm/llvm-project/commit/373c343a77a7afaa07179db1754a52b620dfaf2e you'll see that there were multiple justifications stated for adding the combine to the artifact combiner. One of them was that it was a net compile time benefit and that the call to KB wouldn't be useless work most of the time.

In order to add this to the artifact combiner we'll need similar justification because strictly speaking it doesn't belong here.

**Comment by s-barannikov** [comment]
The justification is just the same. This eliminates redundant instructions after each instruction producing a boolean value (comparisons and long arithmetic). I though it is artifact combiner's job to eliminate the redundant extensions/truncations inserted by the legalizer. Is it not?

I don't have large benchmarks, but on what I have, this patch gives ~1.0% compile time speedup at -O0 and ~1.6% speedup at -O2. .text size is reduced by 2% to 4% depending on optimization level (there _is_ a difference even at -O2 because shl + ashr pair is not currently optimized by the generic combiner).

I didn't measure targets with zero-or-one booleans, but I don't expect the compile time to increase perceptibly because trunc+sext pair is much less likely to occur for them at legalization time. 
I did run `time ninja check-llvm-codegen-amdgpu-globalisel` though (a target with  776 G_SEXT_INREG in GlobalISel tests), and there was no difference in testing time. It might be worth noting that a call to KnownBits is limited by Depth == 2 at -O0.

**Comment by aemerson** [comment]
> The justification is just the same. This eliminates redundant instructions after each instruction producing a boolean value (comparisons and long arithmetic). I though it is artifact combiner's job to eliminate the redundant extensions/truncations inserted by the legalizer. Is it not?
The real reason the artifact combiner is there is because it's actually required in order for our legalization algorithm to terminate. We need to elide unlegalizable artifacts that cancel out in order for it to work at all. Other combines just tag along and we should be careful about adding new ones. That said, if this is purely a mirror for the G_AND one for other boolean format targets then I think it's ok.
> 
> I don't have large benchmarks, but on what I have, this patch gives ~1.0% compile time speedup at -O0 and ~1.6% speedup at -O2. .text size is reduced by 2% to 4% depending on optimization level (there _is_ a difference even at -O2 because shl + ashr pair is not currently optimized by the generic combiner).
> 
> I didn't measure targets with zero-or-one booleans, but I don't expect the compile time to increase perceptibly because trunc+sext pair is much less likely to occur for them at legalization time. I did run `time ninja check-llvm-codegen-amdgpu-globalisel` though (a target with 776 G_SEXT_INREG in GlobalISel tests), and there was no difference in testing time. It might be worth noting that a call to KnownBits is limited by Depth == 2 at -O0.

**Review by aemerson — APPROVED** [review]

**Comment by s-barannikov** [comment]
> The real reason the artifact combiner is there is because it's actually required in order for our legalization algorithm to terminate. We need to elide unlegalizable artifacts that cancel out in order for it to work at all.

I see, thanks.

**Pull request merged**